### PR TITLE
Notebook2 scss

### PR DESCRIPTION
**Pull request by akotlar**
### Stacked PR
Stacks on #5429, adding the scss that was excluded from that PR. I separated them to make the story a bit cleaner.

When #5429 is approved, this PR will contain 3 lines of change to index.html, 5 to notebook.py (to compile scss css), 1 to environment.yml (to include libsass), 1 to gitignore (to ignore the compiled css). The remaining 437 lines are the top-scope styles (margins, buttons that are frequently re-used, etc), the header styles, and styles for the login and notebook pages. We can exclude login page in this PR, but it only adds ~ 12 lines.

Some of the global SCSS we will be able to removed later this week, once we understand which pieces we truly don't need. I can also remove them now if needed, but we won't have a single "styles" PR.

### SCSS vs CSS
I find SCSS preferable to CSS because it makes style scoping and re-use easier. For instance, we may define a variable $margin, to standardize spacing across all pages, and $color to define a font color for all text. SCSS can be written identically to css if the developer is totally unfamiliar or doesn't have time to learn.

My style scoping solution is reminiscent of [BEM](http://getbem.com/introduction/). Each page defines an ID, and any styles that are private to that page are scope that way `#page { ...styles }` in the corresponding `pages/page.scss`. It is lightweight (no additional tooling needed), and should be easy to follow.

cc @cseed

### CODE REVIEW

File: notebook2/notebook/static/styles/variables/margins.scss [hunk 1]
```diff
@@ -0,0 +1 @@
+$margin: 0.875rem;
```
Comment (danking) [inline notebook2/notebook/static/styles/variables/margins.scss]: What's the thinking on defining margin globally in-terms of the root element's em? I think of margins as very UI-element-specfic.

File: notebook2/notebook/static/styles/typography.scss [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    font-size: 20px;
+    letter-spacing: 0.15;
+  }
+  .subtitle-1 {
+    font-size: 16px;
+    letter-spacing: 0.15;
+  }
+  .subtitle-2 {
+    font-size: 14px;
+    font-size: $weight-medium;
```
Comment (danking) [inline notebook2/notebook/static/styles/typography.scss]: this is a weight, no?

File: notebook2/notebook/static/styles/variables/typeography.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+$weight-medium: 500;
+$weight-heavy: 700;
+$weight-light: 300;
```
Comment (danking) [inline notebook2/notebook/static/styles/variables/typeography.scss]: I don't see these used anywhere, except for medium. I guess future PRs will use the other ones? Also, what's the thought on using `500` versus `normal` (defined as `400`)? On my laptop, I see no difference between Open Sans at 500 and Open Sans at 400. Are `normal` and `bold` sufficient for our purposes?

File: notebook2/notebook/static/styles/typography.scss [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    font-size: 24px;
+    letter-spacing: 0;
+  }
+  h6 {
+    font-size: 20px;
+    letter-spacing: 0.15;
+  }
+  .subtitle-1 {
+    font-size: 16px;
+    letter-spacing: 0.15;
```
Comment (danking) [inline notebook2/notebook/static/styles/typography.scss]: Why do all the `letter-spacing`s have no units? It seems like its [supposed to have a unit](https://developer.mozilla.org/en-US/docs/Web/CSS/letter-spacing).

File: notebook2/notebook/static/styles/buttons.scss [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    &.active,
+    &:hover {
+      color: $devil-gray;
+    }
+  }
+  button {
+    outline: none;
+    /* border: none; */
+    /* display: inline-block; */
+    /* width: 20px; */
```
Comment (danking) [inline notebook2/notebook/static/styles/buttons.scss]: Thoughts on these commented out rules?

File: notebook2/notebook/static/styles/buttons.scss [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    padding: 10px;
+    border-radius: 3px;
+    &:focus,
+    &:hover {
+      border-color: $devil-gray;
+    }
+  }
+  .link-button,
+  button {
+    cursor: pointer;
```
Comment (danking) [inline notebook2/notebook/static/styles/buttons.scss]: isn't this the default for anchors and buttons?

File: notebook2/notebook/static/styles/card.scss [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    display: flex;
+    flex-direction: column;
+    .header {
+      text-align: left;
+    }
+    .content {
+      display: flex;
+    }
+  }
+  &.dark {}
```
Comment (danking) [inline notebook2/notebook/static/styles/card.scss]: this seems odd?

File: notebook2/notebook/static/styles/main.scss [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    display: flex;
+    flex-direction: column;
+    padding: $margin;
+    min-height: calc(100vh - 50px * 2);
+    .centered {
+      margin: auto;
+    }
+  }
+  #footer {
+    /* padding: 7px; */
```
Comment (danking) [inline notebook2/notebook/static/styles/main.scss]: commented out rule

File: notebook2/notebook/notebook.py [hunk 9]
```diff
@@ -43,6 +47,12 @@
 app = Flask(__name__)
 sockets = Sockets(app)
 
+scss_path = os.path.join(app.static_folder, 'styles')
+css_path = os.path.join(app.static_folder, 'css')
+os.makedirs(css_path, exist_ok=True)
+
+sass.compile(dirname=(scss_path, css_path), output_style='compressed')
```
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/notebook.py]: In dev context you can adapt this approach to auto-rebuild css when scss changes. If you have a strong opinion on this, and want me to write this into make, I will do that.

File: notebook2/notebook/static/styles/components/header.scss [hunk 10]
```diff
@@ -0,0 +1,68 @@
+@import "../../styles/variables/colors.scss";
+@import "../../styles/variables/margins.scss";
```
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/static/styles/components/header.scss]: Ah yeah, this was copy pasted from app.hail.is, which has a different folder structure (namely header.scss did not have styles as a parent.

File: notebook2/notebook/static/styles/main.scss [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+  box-sizing: inherit;
+}
+
+body {
+  margin: 0px;
+}
+
+/*** https://github.com/tobiasahlin/SpinKit ***/
+
+/*The MIT License (MIT)
```
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/static/styles/main.scss]: From my reading license inclusion is more flexible than this. The spirit is to identify borrowed code. Importing (and keeping the license in the bundle) would do approximately the same thing.

File: notebook2/notebook/static/styles/variables/margins.scss (hunk 1 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/variables/margins.scss]: Every UI element should use a multiple of this value. We need consistent elements across pages.

File: notebook2/notebook/static/styles/typography.scss (hunk 2 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/typography.scss]: Yeah good catch. This style isn't used, should remove.

File: notebook2/notebook/static/styles/typography.scss (hunk 4 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/typography.scss]: Correct, thanks.

File: notebook2/notebook/static/styles/buttons.scss (hunk 5 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/buttons.scss]: Copypasta from app.hail.is, need to clean up.

File: notebook2/notebook/static/styles/buttons.scss (hunk 6 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/buttons.scss]: .link-button can apply to elements other than button.

File: notebook2/notebook/static/styles/card.scss (hunk 7 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/card.scss]: placeholder, will remove

File: notebook2/notebook/static/styles/main.scss (hunk 8 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/main.scss]: thanks!

File: notebook2/notebook/notebook.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (danking) [reply notebook2/notebook/notebook.py]: I don't think the deployed application should compile them at start-up time. It feels to me like the sort of thing that should be resolved when a Docker image is created. A development tool that uses `fswatch` or equivalent to listen for file changes and rebuild sass files seems very reasonable.

[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Are you ok with any of the procedures here: https://sass.github.io/libsass-python/frameworks/flask.html#building-sass-scss-for-each-deployment

File: notebook2/notebook/static/styles/variables/typeography.scss (hunk 3 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/variables/typeography.scss]: It used to be the case that browsers would have different ideas of what “normal” meant. The integer values are precise.

**Comment by akotlar** [comment]
> There's a few direct comments.
> 
> Overall, I must admit there are a lot of rules and it's difficult for me to reason about how they compose. My gut says I should be able to reason about what a page will look like based on the CSS/SASS, but maybe that's not normal?

I tried to make it understandable, but if there’s a way better way let me know!

File: notebook2/notebook/static/styles/variables/typeography.scss (hunk 3 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/variables/typeography.scss]: I've deleted these for now; can add back when neede

File: notebook2/notebook/static/styles/buttons.scss (hunk 6 above)
Comment (akotlar) [reply notebook2/notebook/static/styles/buttons.scss]: @danking, actually, double checked, button does not default to pointer either, at least not in chrome or safari.

**Comment by akotlar** [comment]
@danking I think this is set barring change to lib sass. If possible, I would like to keep the scss compilation in notebook.py, and create an issue to make a better solution as a step 2. I recognize what you want in broad terms, and am happy to do it, and also think what you've mentioned is more complex, requires me to spend additional time to research (auto-reload, not having to retype make scss for every change), and doesn't clearly add value compared to the remaining user-facing work.

**Comment by danking** [comment]
OK, dismiss my review when there's an issue and I'll approve.

**Pull request merged**